src/token/cose/crypto_impl/rustcrypto/sign/ecdsa.rs

```diff
@@ -0,0 +1,205 @@
+    {
+        let digest = Digest::new_with_prefix(payload);
+        let sign_key = Self::cose_ec2_to_ec_private_key::<CRV>(key)?;
+        let (signature, _recid) = <SigningKey<CRV> as DigestSigner<
+            D,
+            (Signature<CRV>, RecoveryId),
+        >>::sign_digest(&sign_key, digest);
+        Ok(signature.to_vec())
+    }
+
+    /// Perform an ECDSA verification operation with the ECDSA hash function given in `oid` for the
+    /// given `payload` and `sig`nature using the provided `key`.
+    fn verify_ecdsa_with_curve<CRV: PrimeCurve + CurveArithmetic + DigestPrimitive>(
+        key: &CoseEc2Key<'_, <Self as CryptoBackend>::Error>,
+        oid: ObjectIdentifier,
+        sig: &[u8],
+        payload: &[u8],
+    ) -> Result<(), CoseCipherError<<Self as CryptoBackend>::Error>>
+    where
+        <CRV as CurveArithmetic>::AffinePoint: VerifyPrimitive<CRV>,
+        <<CRV as ecdsa::elliptic_curve::Curve>::FieldBytesSize as Add>::Output: ArrayLength<u8>,
+        <CRV as ecdsa::elliptic_curve::Curve>::FieldBytesSize: ModulusSize,
+        <CRV as CurveArithmetic>::AffinePoint: FromEncodedPoint<CRV>,
+        <CRV as CurveArithmetic>::AffinePoint: ToEncodedPoint<CRV>,
+    {
+        let sign_key = Self::cose_ec2_to_ec_public_key::<CRV>(key)?;
+        let signature = SignatureWithOid::new(Signature::<CRV>::from_slice(sig)?, oid)?;
+        <VerifyingKey<CRV> as Verifier<SignatureWithOid<CRV>>>::verify(
+            &sign_key, payload, &signature,
+        )
+        .map_err(CoseCipherError::from)
+    }
+
+    /// Convert a public or private COSE EC2 key to its public key RustCrypto representation.
+    fn cose_ec2_to_ec_public_key<CRV: PrimeCurve + CurveArithmetic>(
+        key: &CoseEc2Key<'_, <Self as CryptoBackend>::Error>,
+    ) -> Result<VerifyingKey<CRV>, CoseCipherError<<Self as CryptoBackend>::Error>>
+    where
+        <CRV as ecdsa::elliptic_curve::Curve>::FieldBytesSize: ModulusSize,
+        <CRV as CurveArithmetic>::AffinePoint: FromEncodedPoint<CRV>,
+        <CRV as CurveArithmetic>::AffinePoint: ToEncodedPoint<CRV>,
+    {
+        if key.x.is_none() || (key.y.is_none() && key.sign.is_none()) {
+            // According to the contract provided by the calling COSE library, D must be set, so we
+            // can attempt to reconstruct the public key from the private key.
+            SecretKey::from_slice(key.d.expect(
+                "invalid EC2 key was provided, at least one of the key parameters must be set",
+            ))
+            .map(|sc| VerifyingKey::from(sc.public_key()))
+            .map_err(CoseCipherError::from)
+        } else {
+            // x must be Some here due to the previous condition.
+            let pubkey_coord = if key.y.is_some() {
+                EncodedPoint::<CRV>::from_affine_coordinates(
+                    key.x.unwrap().into(),
+                    key.y.unwrap().into(),
+                    false,
+                )
```

Comment:
Can we use `if let` here to avoid the unwrap on `y`?

Reply:
```suggestion
            let pubkey_coord = if let Some(y) = key.y {
                EncodedPoint::<CRV>::from_affine_coordinates(
                    key.x.unwrap().into(),
                    y.into(),
                    false,
                )
```